x-pack/plugins/observability_solution/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -462,3 +478,178 @@ export function getWaterfall(apiResponse: TraceAPIResponse): IWaterfall {
     orphanTraceItemsCount,
   };
 }
+
+function getChildren({
+  path,
+  waterfall,
+  waterfallItemId,
+}: {
+  waterfallItemId: string;
+  waterfall: IWaterfall;
+  path: {
+    criticalPathSegmentsById: Dictionary<CriticalPathSegment[]>;
+    showCriticalPath: boolean;
+  };
+}) {
+  const children = waterfall.childrenByParentId[waterfallItemId] ?? [];
+  return path.showCriticalPath
+    ? children.filter((child) => path.criticalPathSegmentsById[child.id]?.length)
+    : children;
+}
+
+function buildTree({
+  root,
+  waterfall,
+  maxLevelOpen,
+  path,
+}: {
+  root: IWaterfallNode;
+  waterfall: IWaterfall;
+  maxLevelOpen: number;
+  path: {
+    criticalPathSegmentsById: Dictionary<CriticalPathSegment[]>;
+    showCriticalPath: boolean;
+  };
+}) {
+  const tree = { ...root };
+  const queue: IWaterfallNode[] = [tree];
+
+  for (let queueIndex = 0; queueIndex < queue.length; queueIndex++) {
+    const node = queue[queueIndex];
+
+    const children = getChildren({ path, waterfall, waterfallItemId: node.item.id });
+    // Set childrenToLoad for all nodes enqueued.
+    // this allows lazy loading of child nodes
+    node.childrenToLoad = children.length;
+
+    if (maxLevelOpen > node.level) {
+      children.forEach((child, index) => {
+        const level = node.level + 1;
+
+        const currentNode: IWaterfallNode = {
+          id: btoa(`${node.id}-${child.id}-${index}`),
```

Comment:
I don't know why the waterfall item id isn't unique. The same item can appear many times. This way a unique id is ensured.

Reply:
IIRC, child.id is either span.id if it exists, or transaction.id. Are you using synthtrace? Synthtrace IDs can collide depending on how you index the data (we use incrementing ids to save CPU, but every worker keeps its own counter).

Reply:
Yeah, I'm using synthtrace. the oblt clusters also use synthtrace, right? I suspected that this was a problem with the data generated by synthtrace.

Reply:
yep, those also use synthtrace. So not sure if you want to leave this in, it might lead to other issues?

Reply:
This can be removed, the virtual list sets a unique `key` to the row components.